### ISSUE TITLE
ignore pragmas on typeclass methods applied too much

### ISSUE THREAD

**Issue by ghorn** [ndmitchell/hlint#510]
I have code that looks like

```haskell
instance Arbitrary Foo where
  arbitrary = ...

instance Arbitrary a => Arbitrary (Bar a) where
  {-# ANN arbitrary "HLint: ignore Use <$>" #-}
  arbitrary = ...
```

Hlint is picking up the `ignore` pragma on the `Bar` instance as desired. Unfortunately it is also ignoring `Use <$>` in the `Foo` instance.

**Comment by ndmitchell** [comment]
Hmm, I think that may be just inevitable - HLint doesn't really know where the pragmas are... It collects them all at the beginning into a set.

I guess an alternative question is why don't you want to use `<$>` in the Arbitrary for Bar? If I can improve the hint then that's even better.

**Comment by ghorn** [comment]
I see, it'd be a major structural change to fix that. I've run into it before and used the pattern:
```haskell
instance Arbitrary a => Arbitrary (Bar a) where
  arbitrary = arbitraryBar

{-# ANN arbitraryBar "HLint: ignore Use <$>" #-}
arbitraryBar :: Arbitrary a => Gen (Bar a)
arbitraryBar = ...
```

I did end up just using `<$>` in order to move on, not a big deal.

Since you asked, it was something like:
```haskell
do  x0 <- gen
    x1 <- gen
    x2 <- gen
    x3 <- gen
    ....
    ....
    xNm1 <- gen
    xN <- gen
    return (Bar x0 x1 x2 x3 ...... xNm1 xN)
```
where hlint preferred
```haskell
do  x0 <- gen
    x1 <- gen
    x2 <- gen
    x3 <- gen
    ....
    ....
    xNm1 <- gen
    Bar x0 x1 x2 x3 ...... xNm1 <$> gen
```
I felt it was more readable before and I'd rather not change it because it was a bit legacy anyways.

**Comment by ndmitchell** [comment]
Agreed - I hate it when HLint breaks symmetry - unfortunately I've never been able to figure out what a rule for symmetry looks like (despite being obvious to humans). I can see why you ignore `<$>` in this instance - although I would say the modern idiom for this pattern is `Bar <$> gen <*> gen <*> gen <*> gen ...`
